robots/tank_pid.h

```diff
@@ -0,0 +1,176 @@
+#include <algorithm>
+#include <iostream>
+
+namespace BoBRobotics {
+namespace Robots {
+
+using namespace units::literals;
+
+enum class TankPIDState
+{
+    Invalid,
+    OrientingToGoal,
+    DrivingToGoal,
+    AtGoal
+};
+
+class TankPID
+  : public FSM<TankPIDState>::StateHandler
+{
+    using meter_t = units::length::meter_t;
+    using radian_t = units::angle::radian_t;
+    using second_t = units::time::second_t;
+    using Event = FSM<TankPIDState>::StateHandler::Event;
+
+public:
+    TankPID(Tank &robot, float Kp, float Ki, float Kd,
+            meter_t distanceTolerance = 5_cm,
+            radian_t angleTolerance = 3_deg,
+            radian_t startTurningThreshold = 45_deg, float averageSpeed = 0.5f)
+      : m_StateMachine(this, TankPIDState::Invalid)
+      , m_Robot(robot)
+      , m_DistanceTolerance(distanceTolerance)
+      , m_AngleTolerance(angleTolerance)
+      , m_StartTurningThreshold(startTurningThreshold)
+      , m_Kp(Kp)
+      , m_Ki(Ki)
+      , m_Kd(Kd)
+      , m_AverageSpeed(averageSpeed)
+    {}
+
+    void start(const Vector2<meter_t> &goal)
+    {
+        m_Goal = goal;
+
+        /*
+         * We start off assuming that the heading is way off and that we need to
+         * turn on the spot.
+         */
+        m_StateMachine.transition(TankPIDState::OrientingToGoal);
+    }
+
+    bool driveRobot(const Pose2<meter_t, radian_t> &robotPose)
+    {
+        m_RobotPose = robotPose;
+        const radian_t headingToGoal = units::math::atan2(m_Goal.y() - m_RobotPose.y(),
+                                                          m_Goal.x() - m_RobotPose.x());
+        m_HeadingOffset = circularDistance(headingToGoal, m_RobotPose.yaw());
+
+        // Get state machine to carry out appropriate action
+        m_StateMachine.update();
+
+        // Return true if we're at goal (within m_DistanceTolerance)
+        return m_StateMachine.getCurrentState() == TankPIDState::AtGoal;
+    }
+
+    virtual bool handleEvent(TankPIDState state, Event event) override
+    {
+        using namespace units::math;
+
+        if (event == Event::Update && m_Goal.distance2D(m_RobotPose) < m_DistanceTolerance) {
+            m_StateMachine.transition(TankPIDState::AtGoal);
+            return true;
+        }
+
+        switch (state) {
+        case TankPIDState::OrientingToGoal:
+            if (event == Event::Enter) {
+                std::cout << "Starting turning" << std::endl;
+            } else if (event == Event::Update) {
+                /*
+                 * If m_HeadingOffset is suitably small, we've finished turning
+                 * towards the goal.
+                 */
+                if (abs(m_HeadingOffset) <= m_AngleTolerance) {
+                    // Start driving to the goal in a straight(ish) line
+                    m_StateMachine.transition(TankPIDState::DrivingToGoal);
+                } else if (m_HeadingOffset < 0_deg) {
+                    m_Robot.tank(1.f, -1.f);
```

Comment:
It might be worth slowing this down, especially with small heading offsets to prevent oscillations. Something like ``(min(90, abs(m_HeadingOffset)) / 90) ** 2`` might work nicely or you could just have a threshold, beneath which, it steers slower. That being said, to get passable norbot control/less toppling in the snapshot bot, I gave up on the forementioned approach and just always turned with a really low steering value of ``0.15``.

Reply:
It does work reasonably well as is, although we could slow it down a bit. The danger is we might get stuck on the wobbly floor if it's too slow... This is just for when the robot's heading is really far out, so it does a big turn to get back facing in roughly the right direction.

Reply:
Hah that delicate balance is pretty representative of my attempts at robotting :smile:

Reply:
Thinking about it, you're probably also operating at a higher framerate which probably helps